### PR TITLE
Possibility to use multiple workers to insert requests from Redis to DB

### PR DESCRIPTION
**Pull request by tsteur**
By default still only one worker is used. It is backwards compatible meaning no moving of requests into a different queue is needed etc. More workers can be enabled in the UI under "Admin => Plugin Setting => Number of queue workers".

It basically works the same way as the log importer. Based on the visitorId (fallback IP) we move a request into a queue (there are as many queues as configured workers). Only one worker can work on one queue at a time as we need to make sure all requests within one queue will be inserted in order. This means there is a lock for each queue making sure only one worker is processing a queue. 

Each worker is processing all queues but only one at a time (it is confusing, let me explain this). If we have 6 queues and one worker, this worker will first work on queue 1 and process eg 25 requests (as configured), then move to queue 2 and process 25 requests, then move to queue 3, .... If a queue doesn't need to be processed it will be just skipped and the next queue is checked. If no queue needs processing anymore the worker stops.

If there are more workers running at the same time it works similar just that we check whether a queue is already in process. Meaning if we just inserted 25 requests from queue 2, and queue 3 is in process by another worker, we check whether queue 4 needs to be processed etc. 

Known issues: https://github.com/piwik/piwik/issues/7691

There are many changes in this PR but not much has actually changed. Most changes are new tests and adjusting commands.

Basically, I only introduced a new `Queue\Manager` that manages multiple `Queue`s. The `Queue\Manager` makes sure to only process one `Queue` at a time in `Queue\Manager::lockNext()`. I also added a new class `Queue\Lock` but here I only extracted some existing code into that new class. The `Queue\Processor` can be no longer locked as it did not really make sense to lock a processor. Only a queue can be locked. Otherwise there are more or less no other changes.

### CODE REVIEW

**Comment by tsteur** [comment]
FYI: I just did another test. Replaying logs with and without using queues (with one worker and with multiple workers) into different sites and the result was identical in all sites.

**Comment by tsteur** [comment]
I did a performance test on AWS C3 instance re processing requests from Redis into DB. Please note I inserted into a more or less empty DB. So the more requests there are in our log tables the slower it'll get of course. Ideally we would test it then as well. Also did a test with and without using transactions. We do currently use transactions but it might be worth removing. Even though InooDB locks on row level etc I noticed when profiling multiple workers that time spent in DB goes up quite a lot.

With 2 CPUs and 2 workers: 420 req/s (about 210req/s per worker)

```
******************************************************************************************
  This worker finished queue processing with 209.8req/s (2225 requests in 10.61 seconds)  
******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 212.98req/s (2375 requests in 11.15 seconds)  
******************************************************************************************
```

With 8 CPUs and 8 workers **with** using transactions: 1300req/s (about 160-180req/s per worker)
```
*******************************************************************************************
  This worker finished queue processing with 177.05req/s (6450 requests in 36.43 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 177.49req/s (6975 requests in 39.30 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 186.81req/s (7975 requests in 42.69 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 167.77req/s (5875 requests in 35.02 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 160.93req/s (5300 requests in 32.93 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 168.09req/s (5925 requests in 35.25 seconds)  
*******************************************************************************************

*****************************************************************************************
  This worker finished queue processing with 168.86req/s (325 requests in 1.92 seconds)  
*****************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 159.67req/s (4925 requests in 30.85 seconds)  
*******************************************************************************************
```

With 8 CPUs and 8 workers and **not** using transactions: 1550req/s (about 180-210req/s per worker)

```
*******************************************************************************************
  This worker finished queue processing with 200.17req/s (6300 requests in 31.47 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 181.79req/s (5075 requests in 27.92 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 213.61req/s (8000 requests in 37.45 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 204.66req/s (6925 requests in 33.84 seconds)  
*******************************************************************************************

*****************************************************************************************
  This worker finished queue processing with 171.96req/s (725 requests in 4.22 seconds)  
*****************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 195.58req/s (5850 requests in 29.91 seconds)  
*******************************************************************************************

*******************************************************************************************
  This worker finished queue processing with 181.97req/s (4925 requests in 27.07 seconds)  
*******************************************************************************************

******************************************************************************************
  This worker finished queue processing with 194.6req/s (5975 requests in 30.70 seconds)  
******************************************************************************************
```

**Comment by tsteur** [comment]
Here's a gif of top while inserting the requests. Redis itself is rather bored. 
out
